### PR TITLE
Introduce `IllogicalReactorOperators` bug check

### PR DESCRIPTION
**Pull request by mohamedsamehsalah**
Draft PR to discuss an idea for a bug check observed in a Picnic PR.

I can provide the link to the original bug in private :wink:

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  /** Instantiates a new {@link MonoFilterThen} instance. */
+  public MonoFilterThen() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (MONO_THEN.matches(tree, state)
+        && ASTHelpers.getSymbol(ASTHelpers.getReceiver(tree))
+            .getSimpleName()
+            .toString()
+            .equals("filter")) {
```
Comment (mohamedsamehsalah) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java:47]: Ultimately, we check for all non-terminal operators here.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.errorprone.suppliers.Suppliers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+
+/** TODO: Write Javadoc. */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "You probably meant .filter.when(), right ?",
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java:26]: The `when` method is static, so likely not ;)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    link = BUG_PATTERNS_BASE_URL + "MonoFilterThen",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = LIKELY_ERROR)
+public final class MonoFilterThen extends BugChecker implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Supplier<Type> MONO =
+      Suppliers.typeFromString("reactor.core.publisher.Mono");
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java:34]: Conceptually this check also applies to `Flux`es, right?

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  private static final Matcher<ExpressionTree> MONO_THEN =
+      instanceMethod().onDescendantOf(MONO).named("then");
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java:36]: If we go ahead with this check then there are a few other "element dropping" operations we may want to flag.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  public MonoFilterThen() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (MONO_THEN.matches(tree, state)
+        && ASTHelpers.getSymbol(ASTHelpers.getReceiver(tree))
+            .getSimpleName()
+            .toString()
+            .equals("filter")) {
+      return describeMatch(tree);
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java:48]: My proposal would be to emit a suggested fix that drops the presumed no-op operator (`filter` in this case) :)

**Review by Stephan202 — COMMENTED** [review]
Nice idea! :).

This relates to an idea that came up a few times now: each Reactor operator has some set of properties, and some of those properties are incompatible/non-sensical when combined, so ideally we flag such cases. If we'd have such a framework, then the check proposed here could be incorporated by flagging the composition of operators with the "should be pure" and "drops elements" properties, respectively.

... but that's a decidedly bigger task :sweat_smile: So fleshing out a check as proposed here SGTM!

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java (hunk 4 above)
Comment (mohamedsamehsalah) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoFilterThen.java]: I added something like
```
ImmutableSet.of("then", "thenEmpty", "thenMany")
```

Is there a reference/docs of other operators that can be good candidates for this ?

**Comment by mohamedsamehsalah** [comment]
No plans to progress on this. Need better spec.

Closing for now.

**Pull request closed without merging**